### ISSUE TITLE
Long-running InternalEngine#recoverFromTranslog can block cluster state application

### ISSUE THREAD

**Issue by DaveCTurner** [elastic/elasticsearch#52146]
I saw a 7.5.0 cluster with a node that was failing to apply cluster state updates due to the following chain of locks:

```
"elasticsearch[REDACTED][clusterApplierService#updateTask][T#1]" ... waiting for monitor entry
   java.lang.Thread.State: BLOCKED (on object monitor)
  at org.elasticsearch.indices.cluster.IndicesClusterStateService.applyClusterState(IndicesClusterStateService.java:235)
  - waiting to lock <0x0000000116e271c0> (a org.elasticsearch.indices.cluster.IndicesClusterStateService)
  at org.elasticsearch.cluster.service.ClusterApplierService.lambda$callClusterStateAppliers$5(ClusterApplierService.java:517)
  at org.elasticsearch.cluster.service.ClusterApplierService$$Lambda$4315/0x00000008019c6840.accept(Unknown Source)
  at java.lang.Iterable.forEach(java.base@13.0.1/Iterable.java:75)
  at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateAppliers(ClusterApplierService.java:514)
  at org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:485)
  at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:432)
  at org.elasticsearch.cluster.service.ClusterApplierService.access$100(ClusterApplierService.java:73)
  at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:176)
  at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:703)
  at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:252)
  at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:215)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@13.0.1/ThreadPoolExecutor.java:1128)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@13.0.1/ThreadPoolExecutor.java:628)
  at java.lang.Thread.run(java.base@13.0.1/Thread.java:830)

"elasticsearch[REDACTED][generic][T#9]" ... waiting on condition
   java.lang.Thread.State: WAITING (parking)
  at jdk.internal.misc.Unsafe.park(java.base@13.0.1/Native Method)
  - parking to wait for  <0x0000000127303a98> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
  at java.util.concurrent.locks.LockSupport.park(java.base@13.0.1/LockSupport.java:194)
  at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(java.base@13.0.1/AbstractQueuedSynchronizer.java:885)
  at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(java.base@13.0.1/AbstractQueuedSynchronizer.java:917)
  at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@13.0.1/AbstractQueuedSynchronizer.java:1240)
  at java.util.concurrent.locks.ReentrantReadWriteLock$WriteLock.lock(java.base@13.0.1/ReentrantReadWriteLock.java:959)
  at org.elasticsearch.common.util.concurrent.ReleasableLock.acquire(ReleasableLock.java:55)
  at org.elasticsearch.index.engine.Engine.close(Engine.java:1729)
  at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:104)
  at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:62)
  at org.elasticsearch.index.shard.IndexShard.close(IndexShard.java:1333)
  - locked <0x0000000127371ce8> (a java.lang.Object)
  at org.elasticsearch.index.IndexService.closeShard(IndexService.java:485)
  at org.elasticsearch.index.IndexService.removeShard(IndexService.java:468)
  - locked <0x0000000127f05d68> (a org.elasticsearch.index.IndexService)
  at org.elasticsearch.indices.cluster.IndicesClusterStateService.failAndRemoveShard(IndicesClusterStateService.java:729)
  at org.elasticsearch.indices.cluster.IndicesClusterStateService.handleRecoveryFailure(IndicesClusterStateService.java:721)
  - locked <0x0000000116e271c0> (a org.elasticsearch.indices.cluster.IndicesClusterStateService)
  at org.elasticsearch.indices.cluster.IndicesClusterStateService.access$500(IndicesClusterStateService.java:101)
  at org.elasticsearch.indices.cluster.IndicesClusterStateService$RecoveryListener.onRecoveryFailure(IndicesClusterStateService.java:716)
  at org.elasticsearch.indices.recovery.RecoveryTarget.notifyListener(RecoveryTarget.java:238)
  at org.elasticsearch.indices.recovery.RecoveryTarget.fail(RecoveryTarget.java:225)
  at org.elasticsearch.indices.recovery.RecoveriesCollection.failRecovery(RecoveriesCollection.java:177)
  at org.elasticsearch.indices.recovery.RecoveriesCollection$RecoveryMonitor.doRun(RecoveriesCollection.java:281)
  at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:773)
  at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@13.0.1/ThreadPoolExecutor.java:1128)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@13.0.1/ThreadPoolExecutor.java:628)
  at java.lang.Thread.run(java.base@13.0.1/Thread.java:830)

"elasticsearch[REDACTED][generic][T#17]"
   java.lang.Thread.State: RUNNABLE
  at org.apache.lucene.index.FrozenBufferedUpdates.applyDocValuesUpdates(FrozenBufferedUpdates.java:568)
  at org.apache.lucene.index.FrozenBufferedUpdates.applyDocValuesUpdates(FrozenBufferedUpdates.java:451)
  at org.apache.lucene.index.FrozenBufferedUpdates.apply(FrozenBufferedUpdates.java:421)
  at org.apache.lucene.index.FrozenBufferedUpdates.forceApply(FrozenBufferedUpdates.java:249)
  at org.apache.lucene.index.FrozenBufferedUpdates.tryApply(FrozenBufferedUpdates.java:159)
  at org.apache.lucene.index.IndexWriter.lambda$publishFrozenUpdates$3(IndexWriter.java:2592)
  at org.apache.lucene.index.IndexWriter$$Lambda$5396/0x0000000801c78040.process(Unknown Source)
  at org.apache.lucene.index.IndexWriter.processEvents(IndexWriter.java:5116)
  at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1597)
  at org.apache.lucene.index.IndexWriter.softUpdateDocument(IndexWriter.java:1654)
  at org.elasticsearch.index.engine.InternalEngine.updateDocs(InternalEngine.java:1248)
  at org.elasticsearch.index.engine.InternalEngine.indexIntoLucene(InternalEngine.java:1079)
  at org.elasticsearch.index.engine.InternalEngine.index(InternalEngine.java:923)
  at org.elasticsearch.index.shard.IndexShard.index(IndexShard.java:796)
  at org.elasticsearch.index.shard.IndexShard.applyIndexOperation(IndexShard.java:768)
  at org.elasticsearch.index.shard.IndexShard.applyTranslogOperation(IndexShard.java:1491)
  at org.elasticsearch.index.shard.IndexShard.runTranslogRecovery(IndexShard.java:1522)
  at org.elasticsearch.index.shard.IndexShard.lambda$recoverLocallyUpToGlobalCheckpoint$6(IndexShard.java:1421)
  at org.elasticsearch.index.shard.IndexShard$$Lambda$5353/0x0000000801c58c40.run(Unknown Source)
  at org.elasticsearch.index.engine.InternalEngine.recoverFromTranslogInternal(InternalEngine.java:474)
  at org.elasticsearch.index.engine.InternalEngine.recoverFromTranslog(InternalEngine.java:446)
  at org.elasticsearch.index.engine.InternalEngine.recoverFromTranslog(InternalEngine.java:124)
  at org.elasticsearch.index.shard.IndexShard.recoverLocallyUpToGlobalCheckpoint(IndexShard.java:1427)
  at org.elasticsearch.indices.recovery.PeerRecoveryTargetService.doRecovery(PeerRecoveryTargetService.java:178)
  at org.elasticsearch.indices.recovery.PeerRecoveryTargetService.access$500(PeerRecoveryTargetService.java:79)
  at org.elasticsearch.indices.recovery.PeerRecoveryTargetService$RecoveryRunner.doRun(PeerRecoveryTargetService.java:563)
  at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:773)
  at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@13.0.1/ThreadPoolExecutor.java:1128)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@13.0.1/ThreadPoolExecutor.java:628)
  at java.lang.Thread.run(java.base@13.0.1/Thread.java:830)
```

This third thread holds `Engine#readLock` (not mentioned by `jstack` for some reason) in `InternalEngine.recoverFromTranslog` and holds it until we've finished replaying the local translog. This means the second thread cannot cancel the recovery in a timely fashion during this phase of the recovery.

**Comment by DaveCTurner** [comment]
I think [LUCENE-9228](https://issues.apache.org/jira/browse/LUCENE-9228) explains why the refresh was taking so long. @dnhatn do you think we need to address anything on the Elasticsearch side too, or are we ok to close this?

**Comment by dnhatn** [comment]
Thanks, David, I think we can close this issue.

**Comment by DaveCTurner** [comment]
For the benefit of future readers, the Lucene issue mentioned above can affect a variety of other operations (notably refreshes, and therefore things that block on refreshes like flushes). It only bites if you are repeatedly updating a very small set of document IDs. In most of the cases of this that I've encountered, this was ultimately caused by a client-side bug in how document IDs were generated.

The fix was released in Lucene 8.5 which was included in Elasticsearch 7.7 so you can resolve this issue by upgrading to ≥7.7.0.
